nornir/core/__init__.py

```diff
@@ -79,9 +82,13 @@ def __init__(
         self.configure_logging()
 
         if available_connections is not None:
-            self.available_connections = available_connections
-        else:
-            self.available_connections = connections.available_connections
+            self.data.available_connections = available_connections
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close_all_connections()
```

Comment:
Currently there isn't a method `self.close_all_connections()`, only `self.close_connections()`. It can be fixed either here or where `def close_connections` is defined. I am ok with both naming.
Also, I think it is important to do it safely either here or under the method definition with catch-all try/except block. Since `__exit__` will be called before exit, either graceful or due to an exception, getting another exception during `__exit__` will make it harder to debug. I know it is currently being called using task, but users may use `raise_on_error=True`.